### PR TITLE
Corporate proxy issue prevents connections between MS and GN

### PR DESCRIPTION
**Pull request by shadab-geo**
Updates the HTTP proxy HTTP client dependency to 4.5.13.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -177,12 +177,18 @@
 		</dependency>
 
 		<dependency>
-			<groupId>commons-httpclient</groupId>
-			<artifactId>commons-httpclient</artifactId>
-			<version>3.1</version>
+			<groupId>org.apache.httpcomponents</groupId>
+			<artifactId>httpclient</artifactId>
+			<version>4.5.9</version>
```
Comment (nmco) [inline pom.xml:182]: We should use this version:
https://mvnrepository.com/artifact/org.apache.httpcomponents/httpclient/4.5.13

File: pom.xml [hunk 2]
```diff
@@ -221,6 +227,19 @@
...
+			<artifactId>slf4j-simple</artifactId>
+			<version>1.7.21</version>
+		</dependency>
+
+		<dependency>
+			<groupId>com.github.tomakehurst</groupId>
+			<artifactId>wiremock-jre8-standalone</artifactId>
+			<version>2.32.0</version>
+			<scope>test</scope>
+		</dependency>
```
Comment (nmco) [inline pom.xml:242]: I guess this used to mock the HTTP users? This library comes from GeoTools.

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 3]
```diff
@@ -20,96 +20,261 @@
...
-						+ URLEncoder.encode(fakeLocation, "UTF-8"));
-		final byte[] data = servletOutputStream.baos.toByteArray();
-		Assert.assertNotNull(data);
-		Assert.assertTrue(data.length == 0);
-	}
+
+    private final int localPort = 8080;
+
+    @Rule
+    public WireMockRule wireMockRule =
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:73]: I would recommend making these ports dynamic.

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 4]
```diff
@@ -20,96 +20,261 @@
...
+        Assert.assertNotNull(data);
+        Assert.assertTrue(data.length != 0);
+    }
+
+    /**
+     * Run the local http proxy server to test and pass below Wiremock tests
+     */
+
+    @Test
+    public void testGETUsingWireMock() throws IOException {
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:205]: These tests are indeed more realistic than the full mocked ones, well done. I would recommend creating a new test class named ``HttpProxyIntegrationTests``.

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 5]
```diff
@@ -20,96 +20,261 @@
...
+        proxy.doPost(postRequest, getResponse);
+        verify(getResponse).setStatus(200);
+        final byte[] data = servletOutputStream.baos.toByteArray();
+        Assert.assertNotNull(data);
+        Assert.assertTrue(data.length != 0);
+    }
+
+    /**
+     * Run the local http proxy server to test and pass below Wiremock tests
+     */
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:202]: I would add a better test description. For example in this case something like:
``Validates that a simple HTTP GET requests are correctly handled by the HTTP Proxy.``

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 6]
```diff
@@ -20,96 +20,261 @@
...
-	}
+
+    private final int localPort = 8080;
+
+    @Rule
+    public WireMockRule wireMockRule =
+            new WireMockRule(WireMockConfiguration.options().port(5555));
+
+    @Rule
+    public WireMockRule wireMockProxyRule =
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:77]: I understand that you are expecting the HTTP wire to mock to be running, but it should be started (and stopped) by the test in a transparent way. This should be the way GeoTools is using it.

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 7]
```diff
@@ -20,96 +20,261 @@
...
+        HttpGet httpGet = new HttpGet(proxyURL);
+        try (CloseableHttpClient httpClient = HttpClients.createDefault()) {
+            HttpResponse httpResponse = httpClient.execute(httpGet);
+            Assert.assertEquals(200, httpResponse.getStatusLine().getStatusCode());
+            wireMockRule.verify(getRequestedFor(urlEqualTo("/geostore/users")));
+        }
+    }
+
+    @Test
+    public void testPOSTUsingWireMock() throws IOException {
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:227]: Like above I would add a more meaningful full description for the test.

File: src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java [hunk 8]
```diff
@@ -20,96 +20,261 @@
...
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.message.BasicHeader;
 import org.junit.Assert;
 import org.junit.Before;
+import org.junit.Rule;
 import org.junit.Test;
+import org.mockito.Matchers;
 import org.mockito.Mockito;
 
+import static com.github.tomakehurst.wiremock.client.WireMock.*;
```
Comment (nmco) [inline src/test/java/it/geosolutions/httpproxy/HttpProxyTest.java:61]: What we are missing here is the tests for the new properties (that trigger this update of the library), basically, we will need a "network-proxy" to validate this and set these properties during the test (example values):
```
-Dhttp.proxyHost=proxyhttp.comune.intranet 
-Dhttp.proxyPort=8080 
-Dhttps.proxyHost=proxyhttps.comune.intranet 
-Dhttps.proxyPort=8080 
-Dhttp.nonProxyHosts='localhost|*.comune.intranet|192.168.*'
```

The hard bit here is how to simulate a network proxy for our tests?

My suggestion would be to rely on [WireMock](http://wiremock.org/docs/proxying/) and create three mocked servers:
- The first will act as the network proxy, as described [HERE](http://wiremock.org/docs/proxying/).
- The second (like you are using now) will act as the remote HTTP server we want to reach through the network-proxy.
- The third will act as the internal HTTP server that we don't want to reach via the network-proxy (-Dhttp.nonProxyHosts).

What I'm not sure yet, is out to validate if the requested was or was not routed through the proxy, can you check that @shadab-geo ?

File: pom.xml (hunk 2 above)
Comment (shadab-geo) [reply pom.xml:242]: It is used to test mock endpoints

**Comment by nmco** [comment]
Thx @shadab-geo I will create an issue for the integration tests.

**Pull request merged**